pkg/sql/schemachanger/sctest/cumulative.go

```diff
@@ -37,13 +40,428 @@ import (
+//     statement should be injected. stageEnd accepts the special value
+//     latest which will map to the highest observed stage.
+//   - statements can refer to builtin variable names with a $
+//   - $stageKey - A unique identifier for stages and phases
+//   - $successfulStageCount - Number of stages of the that have been successfully
+//     executed with injections.
+func (m *stageExecStmtMap) parseStageCommon(
+	t *testing.T, d *datadriven.TestData, execType stageExecType,
+) {
+	var key stageKey
+	var schemaChangeErrorRegex *regexp.Regexp
+	stmts := strings.Split(d.Input, ";")
+	require.NotEmpty(t, stmts)
+	// Remove any trailing empty lines.
+	if stmts[len(stmts)-1] == "" {
+		stmts = stmts[0 : len(stmts)-1]
+	}
+	for _, cmdArgs := range d.CmdArgs {
+		switch cmdArgs.Key {
+		case "phase":
+			found := false
+			for i := scop.EarliestPhase; i <= scop.LatestPhase; i++ {
+				if cmdArgs.Vals[0] == i.String() {
+					key.phase = i
+					found = true
+					break
+				}
+			}
+			require.Truef(t, found, "invalid phase name %s", cmdArgs.Key)
+			if !found {
+				panic("phase not mapped")
+			}
+		case "stage":
+			// Detect ranges, otherwise we are looking at single value.
+			if strings.Contains(cmdArgs.Vals[0], ":") {
+				rangeVals := strings.Split(cmdArgs.Vals[0], ":")
+				key.minOrdinal = 1
+				key.maxOrdinal = stageKeyOrdinalLatest
+				if len(rangeVals) >= 1 && len(rangeVals[0]) > 0 {
+					ordinal, err := strconv.Atoi(rangeVals[0])
+					require.Greater(t, ordinal, 0, "minimum ordinal is zero")
+					require.NoError(t, err)
+					key.minOrdinal = ordinal
+				}
+				if len(rangeVals) == 2 && len(rangeVals[1]) > 0 {
+					ordinal, err := strconv.Atoi(rangeVals[1])
+					require.Greater(t, ordinal, 0, "minimum ordinal is zero")
+					require.NoError(t, err)
+					require.GreaterOrEqualf(t, key.maxOrdinal, key.minOrdinal, "ordinals range is invalid")
+					key.maxOrdinal = ordinal
+				}
+			} else {
+				ordinal, err := strconv.Atoi(cmdArgs.Vals[0])
+				require.Greater(t, ordinal, 0, "minimum ordinal is zero")
+				require.NoError(t, err)
+				key.minOrdinal = ordinal
+				key.maxOrdinal = ordinal
+			}
+		case "schemaChangeExecError":
+			schemaChangeErrorRegex = regexp.MustCompile(strings.Join(cmdArgs.Vals, " "))
+		default:
+			require.Failf(t, "unknown key encountered", "key was %s", cmdArgs.Key)
+		}
+	}
+	entry := stageKeyEntry{
+		stageKey: key,
+		stmt: &stageExecStmt{
+			execType:               execType,
+			stmts:                  stmts,
+			observedOutput:         "",
+			expectedOutput:         d.Expected,
+			schemaChangeErrorRegex: schemaChangeErrorRegex,
+		},
+	}
+	m.entries = append(m.entries, entry)
+	m.rewriteMap[d.Pos] = entry.stmt
+}
+
+// GetExpectedOutputForPos returns the expected output for a given line,
+// in rewrite mode this is updated.
+func (m *stageExecStmtMap) GetExpectedOutputForPos(pos string) string {
+	return m.rewriteMap[pos].expectedOutput
```

Comment:
I didn't understand this part. I wanted to figure out how "--rewrite" works: If I now run the EndToEndSideEffect test with a `--rewrite` flag, what will happen? You said we will ignore those commands in EndToEndSideEffect tests, then shouldn't that nothing happens?

Reply:
So, the way the rewrite works is that on the first execution of the injection tests we will gather the expected output mapped by filename + line number. On the second output, this stored output will be returned so that the data driven test can do a rewrite.